### ISSUE TITLE
build error at commit 55f70f615226b2585d213fa15dfa28d6811c8d14 

### ISSUE THREAD

**Issue by ghost** [SDL-Hercules-390/hyperion#674]
environment 
Apple macbook pro M3 
Apple sonoma
Apple clang version 15.0.0 (clang-1500.3.9.4)

( worked until saturday )

```
[enrico@enrico-mbp build]$git log -n 1     
commit 55f70f615226b2585d213fa15dfa28d6811c8d14 (HEAD -> develop, origin/develop)
Author: Fish (David B. Trout) <fish@infidels.org>
Date:   Sun Jul 21 19:30:26 2024 -0700

    Report unavailable host instructions
[enrico@enrico-mbp build]$gmake
gmake  all-recursive
Making all in m4
gmake[2]: Nothing to be done for 'all'.
Making all in util
gmake[2]: Nothing to be done for 'all'.
Making all in html
gmake[2]: Nothing to be done for 'all'.
Making all in man
gmake[2]: Nothing to be done for 'all'.
Making all in .
  CC       impl.lo
../impl.c:876:13: error: no member named 'v' in 'QW'
        mm1.v = _mm_setzero_si128();
        ~~~ ^
../impl.c:876:17: error: call to undeclared function '_mm_setzero_si128'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
        mm1.v = _mm_setzero_si128();
                ^
../impl.c:878:13: error: no member named 'v' in 'QW'
        mm2.v = _mm_setzero_si128();
        ~~~ ^
../impl.c:881:13: error: no member named 'v' in 'QW'
        acc.v = _mm_clmulepi64_si128 ( mm1.v, mm2.v, 0);
        ~~~ ^
../impl.c:881:17: error: call to undeclared function '_mm_clmulepi64_si128'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
        acc.v = _mm_clmulepi64_si128 ( mm1.v, mm2.v, 0);
                ^
../impl.c:881:44: error: no member named 'v' in 'QW'
        acc.v = _mm_clmulepi64_si128 ( mm1.v, mm2.v, 0);
                                       ~~~ ^
../impl.c:881:51: error: no member named 'v' in 'QW'
        acc.v = _mm_clmulepi64_si128 ( mm1.v, mm2.v, 0);
                                              ~~~ ^
../impl.c:1427:5: warning: ../impl.c(1427) : warning : This build of Hercules will only have partial TCP keepalive support [-W#pragma-messages]
    WARNING("This build of Hercules will only have partial TCP keepalive support")
    ^
../ccfixme.h:102:33: note: expanded from macro 'WARNING'
  #define WARNING( _str )       QPRAGMA( message(  WARN_LINE  _str ))
                                ^
../ccfixme.h:64:33: note: expanded from macro 'QPRAGMA'
  #define QPRAGMA( x )          _Pragma( #x )
                                ^
<scratch space>:64:2: note: expanded from here
 message( WARN_LINE "This build of Hercules will only have partial TCP keepalive support" )
 ^
1 warning and 7 errors generated.
gmake[2]: *** [Makefile:2637: impl.lo] Error 1
gmake[1]: *** [Makefile:2687: all-recursive] Error 1
gmake: *** [Makefile:2006: all] Error 2
```

e

**Comment by Fish-Git** [comment]
The cause of the problem appears to be  here:

https://github.com/SDL-Hercules-390/hyperion/blob/62cbbfcf2ac71ee3f5692c465d07ecf0d545f877/esa390.h#L78-L80

Apparently Apple's clang isn't setting `__SSE2__`? Enrico: can you confirm this?

Have you tried adding the sse4.1 and sse4.2 compiler options:

`--enable-optimization="............ -msse4.1 -msse4.2 -mpclmul" `

to your `./configure` command's `--enable-optimization=` option? Maybe we need to add those options to our default auto-optimization build options in our `configure.ac` script?

Bill? Does it build okay (or do Enrico's errors occur?) when you try it with your Hercules Helper?

Help! &nbsp;`:(`

**Comment by Fish-Git** [comment]
So Apple doesn't support _any_ SIMD intrinsics?? Or is its SIMD intrinsics support simply _different??_

I guess we should maybe add some tests for `__APPLE__` to a few of our `#if...` statements to temporarily (or _permanently?_) work around the problem until we get this straightened out?

Or _**better**_, add some new `#defines` to our `hostopts.h` member, and test for them instead! &nbsp;_(Refer to the "*BUG*:  FIX PORTABILITY ISSUES that have crept in over time" section of our `_TODO.txt` member!)_

**Comment by JamesWekel** [comment]
Fish,

The Apple is an M3 processor, ARM64 not X86_86. 

The linux version of "is_PCLMULQDQ_available()" just needs to be conditional on predefined macro X84_64 being defined.

Jim

**Comment by atncsj6h** [comment]
the SIMD things are supported  for the x86_64 architecture

the Mx family of apple processors are arm64/aarch64 processors
the equivalent arm functions are supported by the NEON intrinsics

**Comment by JamesWekel** [comment]
Fish,

My apologies, the Linux version of "is_PCLMULQDQ_available()" needs to be conditional equivalent to the gcc/clang inclusion of intrinsic hstdinc.h:

```
#if defined( __GNUC__) && defined( __SSE2__ ) && ( __SSE2__ == 1 )
  #include <x86intrin.h>
  #define _GCC_SSE2_
#endif
```
On the Apple M3, Intel intrinsics are not included as __SSE2__ can only be defined on a X86_64 when SSE2 is detected. Just being conditional on X86_64 is not sufficient.

Most of the errors are because the intrinsic, <x86intrin.h>, is not included. 

The Apple M3 which uses Apple's aarch64 version of ARM64 has equivalent SIMD features, but implemented with different instructions. By introducing Intel intrinsic usage, we are currently, just optimizing for X64 processors. In order to optimize for Apple M3 (and other 
 aarch64) processors, we would have to include <arm_neon.h> with completely difference intrinsics and its associated learning curve. This leads to a discussion regarding hardware platform optimization: what platforms, how to test for hardware features, the optimization method, ....  The default is always portable C code as currently for the Apple M3.

Hope this helps,
Jim

**Comment by wrljet** [comment]
> 
> Bill? Does it build okay (or do Enrico's errors occur?) when you try it with your Hercules Helper?

Fails the same as Enrico's.  But I guess this is in-hand now.

Bill

**Comment by Fish-Git** [comment]
Enrico, Bill,

Please apply the following simple patch, and let me know if it fixes your problem. If it does, I'll go ahead and commit it.


* [fix674.zip](https://github.com/user-attachments/files/16365471/fix674.zip)

```diff
--- hyperion-1/impl.c	2024-07-21 19:30:45.648700900 -0700
+++ hyperion-0/impl.c	2024-07-24 10:06:59.047791100 -0700
@@ -832,7 +832,7 @@
 
 #else // !defined( _MSVC_ ), i.e. Linux
 
-#if defined( HAVE_SIGNAL_HANDLING )
+#if defined( _GCC_SSE2_ ) && defined( HAVE_SIGNAL_HANDLING )
 
 static struct sigaction  sa_CRASH   = {0};
 static struct sigaction  sa_SIGILL  = {0};
@@ -898,15 +898,15 @@
   #pragma GCC pop_options
 #endif
 
-#else // !defined( HAVE_SIGNAL_HANDLING )
+#else // !defined( _GCC_SSE2_ ) || !defined( HAVE_SIGNAL_HANDLING )
 
-/* No way to know without signal handling, so play it safe! */
+/* No way to know without SSE2 and signal handling, so play it safe! */
 static void is_PCLMULQDQ_available()
 {
     sysblk.have_PCLMULQDQ = false;  /* (safest default) */
 }
 
-#endif // defined( HAVE_SIGNAL_HANDLING )
+#endif // defined( _GCC_SSE2_ ) && defined( HAVE_SIGNAL_HANDLING )
 #endif /* Windows or Linux */
 
 /* Check if various host instructions are available or not */
```

**Comment by atncsj6h** [comment]
both patches worked 
 
since right now I am the only one affected
( I only included the arm_neon_header without using the intrinsics )
the issue is not urgent,  I  would prefer to delay the yes_it_works  for fix675.patch 
after having implemented  and tested at least a couple of the neon intrinsics 

RANT ON
I just hate the #undef #define logic 
I keep forgetting what in hell was undeffed 
and to what in the adjacent hell I am referring to when using float32_t and sfloat32_t  
RANT OFF

anyway ... 

IMO to provide an easier path for enhancements 
it would be better to separate things using the arch macros provided by the compiler  

right now I am trying to implement the neon intrinsics and the current #if logic is unreasonable ...
I have to cleanup everything by hand

something along the lines of

# if defined( USE_INTRINSICS) && defined( __x86_64__ )
 here insert the current intrinsics check and implementation logic
 which I will leave alone
#elif defined( USE_INTRINSICS) &&  defined (__aarch64__ )
 insert the neon/sse2neon  intrinsics interface logic
 my playground for neon intrinsics interface 
#else  
 do not use intrinsics
#endif

might provide better portability with a clear boundaries definition


e

**Comment by Fish-Git** [comment]
> MO to provide an easier path for enhancements
> it would be better to separate things using the arch macros provided by the compiler

I agree wholeheartedly! I was lazy with my test. I just wanted to see if it would work for you. When I do the commit, I will protect it with the your suggested ARM `#if...` statements.

**Comment by Fish-Git** [comment]
Fixed by commit 96705ba1532b9f1c61d04cf33fb6cece7dc1c583.

**Comment by wrljet** [comment]
Builds for me without trouble on macOS Sonoma 14.5, Apple M1 CPU, Clang 15.0.0.
